### ISSUE TITLE
The `onCancel` handler was attached after the promise settled

### ISSUE THREAD

**Issue by myfreeer** [sindresorhus/got#1187]
#### Describe the bug

 - Node.js version: 12.16.2 x64 (LTS)
 - OS & version: Windows 10 LTSC x64
 - got version: [v11.0.2](https://github.com/sindresorhus/got/releases/tag/v11.0.2)

Using `Promise API` with socket timeout and retry, this does not reproduce on Node.js 14.0.0.

#### Actual behavior
Promise rejected before retry, and stderr printed message like below:
```
(node:9792) UnhandledPromiseRejectionWarning: Error: The `onCancel` handler was attached after the promise settled.
    at onCancel (D:\TEMP\test-case\node_modules\p-cancelable\index.js:46:12)
    at makeRequest (D:\TEMP\test-case\node_modules\got\dist\source\as-promise\index.js:39:13)
    at Timeout.retry [as _onTimeout] (D:\TEMP\test-case\node_modules\got\dist\source\as-promise\index.js:182:25)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7)
(node:9792) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:9792) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

Edited: some logs showing the timestamp.
```
1587995291087 response event fired
1587995291088 before await getStream.buffer(request)
1587995291117 error event fired
1587995291118 catch block after await getStream.buffer(request)
1587995291119 promise rejected in catch block
1587995292181 retry
```

#### Expected behavior
Retryed.

#### Code to reproduce

```js
const http = require('http');

const got = require('got').extend({
  timeout: {
    socket: 20
  },
  retry: {
    limit: 5
  }
});

let requested = false;

let server = http.createServer(((req, res) => {
  if (requested) {
    res.writeHead(200, {
      'Content-Length': 8
    });
    res.write('ok123456');
    return;
  }
  req.socket.pause();
  setTimeout(() => {
    res.writeHead(200, {
      'Content-Length': 8
    });
    res.write('ok');
  }, 10);
  setTimeout(() => {
      res.write('123');
  }, 20);
  setTimeout(() => {
      requested = true;
      res.write('456');
      res.end();
  }, 41);
}));


server.listen(45678, 'localhost', async () => {
  try {
    let r = await got('http://localhost:45678/');
    console.log(r && r.body);
  } catch (e) {
    console.log('promise rejected', e);
  }
  setTimeout(() => server.close(), 400);
});
```

#### Checklist

- [x] I have read the documentation.
- [x] I have tried my code with the latest version of Node.js and Got.

**Comment by Watanuki-Kimihiro** [comment]
The same error.

**Comment by szmarczak** [comment]
Simplest repro:

```js
const http = require('http');

const got = require('.').extend({
	timeout: 20,
	retry: {
		limit: 1
	}
});

const server = http.createServer((_request, response) => {
	response.write('ok');
});


server.listen(45678, 'localhost', async () => {
	try {
		const {body} = await got('http://localhost:45678/');
		console.log(body);
	} catch (error) {
		console.log('promise rejected', error);
	}
});
```

**Comment by szmarczak** [comment]
Wow, in fact there are 3 unhandled errors when running inside tests.

**Comment by cjroebuck** [comment]
Still getting this error on 11.3.0

**Comment by szmarczak** [comment]
@cjroebuck Can you post reproducible code please?

**Comment by cjroebuck** [comment]
Hmm. It's happening about once per 12-24 hours, in production, Sorry I'm not sure how to repro. 

These are the options I'm using though:

```ts
  const gotOpts = {
    form: query,
    timeout: 10000,
    throwHttpErrors: false,
    retry: {
      limit: 2,
      calculateDelay: ({ attemptCount, retryOptions, error, computedValue }) => {
        if (computedValue) {
          // let retryIn = 1000 * Math.pow(2, attemptCount) + Math.random() * 100;
          logger.warn("Got an error from engine, retry in", computedValue, longRequestId, { attemptCount, error: error.message });
          return computedValue;
        }
        logger.info("Do not retry", error.message, longRequestId)
        return 0;
      },
      methods: ['POST'] as Method[]
    }
  };
let response = await got.post(someUrl, gotOpts);
```
stack trace:

```text
"Error: The `onCancel` handler was attached after the promise settled. 
at onCancel (/usr/src/app/node_modules/p-cancelable/index.js:46:12)
at makeRequest (/usr/src/app/node_modules/got/dist/source/as-promise/index.js:39:13)
at Timeout.retry [as _onTimeout] (/usr/src/app/node_modules/got/dist/source/as-promise/index.js:178:25)
at ontimeout (timers.js:436:11)
at tryOnTimeout (timers.js:300:5)
at listOnTimeout (timers.js:263:5)
at Timer.processTimers (timers.js:223:10)
```

**Comment by szmarczak** [comment]
@cjroebuck Thanks. What Node.js version are you using?

**Comment by szmarczak** [comment]
Seems like the `response` event is emitted after `request.destroy()` was called, which is pretty werid.

**Comment by szmarczak** [comment]
Or

https://github.com/sindresorhus/got/blob/7582d916db72f7bab2b2c162d37cfcfde08bbba7/source/as-promise/index.ts#L79

resolves when executing `request.destroy()`.

**Comment by cjroebuck** [comment]
@szmarczak Using node 10. Under what conditions is `request.destroy()` called? I'm running in Kubernetes and the network request is from one pod to a service. It's possible that either the pod sending or receiving the request could get killed at any moment and leave the connection in an unfinished state. Since it's only happening  a few times per day, this is most likely whats causing the weird sequence of events.

**Comment by szmarczak** [comment]
> Using node 10. Under what conditions is request.destroy() called?

When there's a timeout. I think I know how to reproduce this. Thanks for the info.

**Comment by szmarczak** [comment]
Either the `response` event emits `IncomingMessage` which has got `complete` property set to `true` or the timeout occurs right after the data has been read.

> Seems like the response event is emitted after request.destroy() was called, which is pretty weird.

That's not possible, because there's `request.off('response', onResponse);` a few lines before.

**Comment by szmarczak** [comment]
Hmm... The only possibility I see is when `error` is emitted right after the response `end` event.

But it may be related to your POST request. Retrying on POST is quite unsafe. What does your `query` variable look like?

**Comment by szmarczak** [comment]
@cjroebuck Is the error an unhandled promise or you caught it in the `catch` clause?

**Comment by cjroebuck** [comment]
Unhandled rejection.

**Comment by cjroebuck** [comment]
The query variable is a query string object, next time it happens I'll try and grab the contents. I have updated to node 14 in prod though based on what myfreer said initially.

**Comment by szmarczak** [comment]
Can be easily reproduced on master with

```js
const got = require('.');
const http = require('http');

const server = http.createServer((request, response) => {
	// If this was just `.write('asdf')`, it wouldn't throw.
	response.end();
});

server.listen(8080, async () => {
    try {
		await got('http://localhost:8080', {
			timeout: 100,
			retry: 1
		}).on('request', request => {
			const timestamp = Date.now();

			request.once('error', () => {
				console.log(`${timestamp} Received error`);
			});

			request.once('response', response => {
				console.log(`${timestamp} Got response`);

				response.on('readable', () => {
					console.log(`${timestamp} Response is readable`, response.complete);

					// setImmediate works too
					setTimeout(() => {
						const error = new Error('Blah blah blah');
						error.code = 'ETIMEDOUT';
						request.destroy(error);
					});
				});

				response.once('end', () => {
					console.log(`${timestamp} Response ended`);
				});
			});
		});
    } catch (error) {
        console.log(error.message);
	}
	console.log('the end');
});
```

It gives me:

```
1593728001900 Got response
1593728001900 Response is readable true
1593728001900 Response ended
the end
1593728001900 Received error
(node:16624) UnhandledPromiseRejectionWarning: Error: The `onCancel` handler was attached after the promise settled.
    at onCancel (/home/szm/Desktop/got/node_modules/p-cancelable/index.js:46:12)
    at makeRequest (/home/szm/Desktop/got/dist/source/as-promise/index.js:39:13)
    at Timeout.retry [as _onTimeout] (/home/szm/Desktop/got/dist/source/as-promise/index.js:181:25)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7)
```

IOW the timeout occurs when the final chunk arrived, but the `end` event hasn't been emitted yet.

**Comment by szmarczak** [comment]
@cjroebuck Can you replace

```js
let response = await got.post(someUrl, gotOpts);
```

with

```js
let response = await got.post(someUrl, gotOpts).on('request', request => {
	const timestamp = Date.now();

	request.once('error', () => {
		console.log(`${timestamp} Received error`);
	});

	request.once('response', response => {
		console.log(`${timestamp} Got response`);

		response.on('readable', () => {
			console.log(`${timestamp} Response is readable`, response.complete);
		});

		response.once('end', () => {
			console.log(`${timestamp} Response ended`);
		});
	});
});
```

If you could send me the logs, it would be great.

**Comment by szmarczak** [comment]
@cjroebuck Does setting `retry.limit` to `1` fix the issue?

**Comment by szmarczak** [comment]
@cjroebuck What's the status of the `let response = await got.post(someUrl, gotOpts);` promise? Is it resolved or rejected? If rejected, what error does it give?

**Comment by szmarczak** [comment]
You can also make `calculateDelay` log the error.

**Comment by szmarczak** [comment]
Related: #1308

**Comment by Stono** [comment]
Hey, i'm seeing this on nodejs14.

I rolled back versions of got and discovered that `11.1.4` is fine, so its > than that version which has the problem

**Comment by szmarczak** [comment]
@Stono Thanks for the feedback! Can you apply https://github.com/sindresorhus/got/issues/1187#issuecomment-653242747 to the promise and send me the logs please?

**Comment by Stono** [comment]
Apologies @szmarczak - my issue was slightly different to this onCancel one, i've raised a separate issue: https://github.com/sindresorhus/got/issues/1343

**Comment by szmarczak** [comment]
This should be fixed in `11.4.0`.

**Comment by foxundermoon** [comment]
i meet the issue on version: `11.8.5 `

> This should be fixed in `11.4.0`.

**Comment by szmarczak** [comment]
That's probably #1489
